Src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs

```diff
@@ -337,5 +338,137 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
                     _ => continuationFunction(t), cancellationToken, TaskContinuationOptions.None, scheduler),
                 cancellationToken, taskContinuationOptions, scheduler).Unwrap();
         }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        public static Task<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
```

Comment:
I'm not sure I understand the concern, @jasonmalinowski. Are you saying the cancellationToken should be sent to the task when it's created and it should respond quickly enough to cancellation that you don't need this? If you are, I agree. However I've found this extension method useful for those tasks you cannot guarantee will respond in an acceptable time period yet you need to ensure the caller isn't held at its mercy.
But if you'd rather not have this extension method, and you're OK with just calling WhenAll without cancellation, I can avoid adding this to your code base. But personally I think this extension method is more generally useful than a WhenAll specific extension method would be.